### PR TITLE
Task/json serialization

### PR DESCRIPTION
**Pull request by bgulanowski**
# Purpose

This extends the functionality of the Core Data managed object description classes with support for JSON serialization driven by metadata encoded in a managed object model file. These extensions work independently of a Core Data persistence stack: no database or context required.

The extensions support the mapping of property names and values, including attributes and relationships, into JSON and back. Names are mapped using a convention for converting camelCase in Cocoa to snake_case in JSON, as understood by a Ruby On Rails REST API.

The mapping of both names and values is fully customizable by specifying alternative values for names, and transformers for values, in the user info dictionaries associated with properties in the managed object model file.

Continues from https://github.com/Shopify/mobile-buy-sdk-ios/pull/130

## TODO

- [ ] NSEntityDescription additions tests
- [ ] NSPropertyDescription (& NSAttributeDescription & NSRelationshipDescription) Additions tests

@davidmuzi @gabrieloc @dbart01

### CODE REVIEW

**Comment by bgulanowski** [comment]
- [ ] write tests for new categories

**Comment by dbart01** [comment]
Compiles, LGTM 🎩

**Comment by bgulanowski** [comment]
This PR has been updated with tests and some other changes. Please provide final review and verify test coverage @dbart01 @davidmuzi

File: Mobile Buy SDK/Mobile Buy SDK/Additions/NSPropertyDescription+BUYAdditions.m [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+}
+
+// model -> JSON
+- (id)buy_JSONForObject:(NSObject<BUYObject> *)object
+{
+	// JSON generation for a relationship depends on the rules defined in the model.
+	// The model can explicitly specify using an `id` encoding.
+	// Alternately, if the relationship is compatible, encode the entire object.
+	// We do not encode related objects unless three conditions are satisfied:
+	// 1. the relationship is not many-to-many
```
Comment (davidmuzi) [inline Mobile Buy SDK/Mobile Buy SDK/Additions/NSPropertyDescription+BUYAdditions.m]: how/where is this checked?

**Comment by davidmuzi** [comment]
One comment.  Branch needs updating.

 otherwise tested and reviewed 👍

File: Mobile Buy SDK/Mobile Buy SDK/Additions/NSPropertyDescription+BUYAdditions.m (hunk 1 above)
Comment (bgulanowski) [inline Mobile Buy SDK/Mobile Buy SDK/Additions/NSPropertyDescription+BUYAdditions.m]: In `-buy_JSONForValue:`. These comments would probably be more appropriate in that method, which invokes this one (for to-one relationships) and others.

**Comment by bgulanowski** [comment]
@davidmuzi I discovered and fixed a bug and added a new test for it, some other small changes, and moved that comment.

**Pull request merged**